mmengine/analysis/complexity_analysis.py

```diff
@@ -0,0 +1,381 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+import typing
+from collections import defaultdict
+from typing import Any, Counter, DefaultDict, Dict, Optional, Tuple, Union
+
+import tabulate
+import torch.nn as nn
+from torch import Tensor
+
+from .jit_analysis import JitModelAnalysis
+from .jit_handles import (Handle, addmm_flop_jit, batchnorm_flop_jit,
+                          bmm_flop_jit, conv_flop_jit, einsum_flop_jit,
+                          elementwise_flop_counter, generic_activation_jit,
+                          linear_flop_jit, matmul_flop_jit, norm_flop_counter)
+
+# A dictionary that maps supported operations to their flop count jit handles.
+_DEFAULT_SUPPORTED_FLOP_OPS: Dict[str, Handle] = {
+    'aten::addmm': addmm_flop_jit,
+    'aten::bmm': bmm_flop_jit,
+    'aten::_convolution': conv_flop_jit,
+    'aten::einsum': einsum_flop_jit,
+    'aten::matmul': matmul_flop_jit,
+    'aten::mm': matmul_flop_jit,
+    'aten::linear': linear_flop_jit,
+    # You might want to ignore BN flops due to inference-time fusion.
+    # Use `set_op_handle("aten::batch_norm", None)
+    'aten::batch_norm': batchnorm_flop_jit,
+    'aten::group_norm': norm_flop_counter(2),
+    'aten::layer_norm': norm_flop_counter(2),
+    'aten::instance_norm': norm_flop_counter(1),
+    'aten::upsample_nearest2d': elementwise_flop_counter(0, 1),
+    'aten::upsample_bilinear2d': elementwise_flop_counter(0, 4),
+    'aten::adaptive_avg_pool2d': elementwise_flop_counter(1, 0),
+    'aten::grid_sampler': elementwise_flop_counter(0, 4),  # assume bilinear
+}
+
+# A dictionary that maps supported operations to
+# their activation count handles.
+_DEFAULT_SUPPORTED_ACT_OPS: Dict[str, Handle] = {
+    'aten::_convolution': generic_activation_jit('conv'),
+    'aten::addmm': generic_activation_jit(),
+    'aten::bmm': generic_activation_jit(),
+    'aten::einsum': generic_activation_jit(),
+    'aten::matmul': generic_activation_jit(),
+    'aten::linear': generic_activation_jit(),
+}
```

Comment:
Why do we only need to calculate those operations which are a subset of `_DEFAULT_SUPPORTED_FLOP_OPS`?

Reply:
Currently, fvcore only supports the activation computation over the common operators. Also many advanced layers can be parsed with these basic operators(e.g. MultiHeadAttention consists of matmul, einsum, linear, etc.)